expression/builtin_encryption.go

```diff
@@ -548,97 +506,169 @@ func (b *builtinSHA2Sig) eval(row []types.Datum) (d types.Datum, err error) {
+	binary.LittleEndian.PutUint32(buffer, uint32(len(str)))
+	copy(buffer[4:], compressed)
+
+	if shouldAppendSuffix {
+		buffer[len(buffer)-1] = '.'
+	}
+
+	return string(buffer), false, nil
+}
+
+type uncompressFunctionClass struct {
+	baseFunctionClass
+}
+
+func (c *uncompressFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpString, tpString)
 	if err != nil {
-		return d, errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
-	d.SetBytes(bytes)
-	return d, nil
+	bf.tp.Flen = mysql.MaxBlobWidth
+	types.SetBinChsClnFlag(bf.tp)
+	sig := &builtinUncompressSig{baseStringBuiltinFunc{bf}}
+	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
+}
+
+type builtinUncompressSig struct {
+	baseStringBuiltinFunc
+}
+
+// evalString evals UNCOMPRESS(compressed_string).
+// See https://dev.mysql.com/doc/refman/5.7/en/encryption-functions.html#function_uncompress
+func (b *builtinUncompressSig) evalString(row []types.Datum) (string, bool, error) {
+	payload, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
+	if isNull || err != nil {
+		return "", true, errors.Trace(err)
+	}
+	if len(payload) == 0 {
+		return "", false, nil
+	}
+	if len(payload) <= 4 {
+		// corrupted
+		return "", true, nil
+	}
+	bytes, err := inflate([]byte(payload[4:]))
+	if err != nil {
+		return "", true, nil
+	}
+	return string(bytes), false, nil
 }
 
 type uncompressedLengthFunctionClass struct {
 	baseFunctionClass
 }
 
 func (c *uncompressedLengthFunctionClass) getFunction(args []Expression, ctx context.Context) (builtinFunc, error) {
-	sig := &builtinUncompressedLengthSig{newBaseBuiltinFunc(args, ctx)}
+	bf, err := newBaseBuiltinFuncWithTp(args, ctx, tpInt, tpString)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	sig := &builtinUncompressedLengthSig{baseIntBuiltinFunc{bf}}
 	return sig.setSelf(sig), errors.Trace(c.verifyArgs(args))
 }
 
 type builtinUncompressedLengthSig struct {
-	baseBuiltinFunc
+	baseIntBuiltinFunc
 }
 
-// eval evals a builtinUncompressedLengthSig.
+// evalInt evals UNCOMPRESSED_LENGTH(str).
 // See https://dev.mysql.com/doc/refman/5.7/en/encryption-functions.html#function_uncompressed-length
-func (b *builtinUncompressedLengthSig) eval(row []types.Datum) (d types.Datum, err error) {
-	args, err := b.evalArgs(row)
+func (b *builtinUncompressedLengthSig) evalInt(row []types.Datum) (int64, bool, error) {
+	payload, isNull, err := b.args[0].EvalString(row, b.ctx.GetSessionVars().StmtCtx)
 	if err != nil {
-		return d, errors.Trace(err)
+		return 0, true, errors.Trace(err)
 	}
-	arg := args[0]
-	if arg.IsNull() {
-		return d, nil
+	if isNull {
+		return 0, true, nil
 	}
-	compressStr, err := arg.ToBytes()
-	if err != nil {
-		return d, errors.Trace(err)
+	if len(payload) == 0 {
+		return 0, false, nil
 	}
-	if len(compressStr) == 0 {
-		d.SetInt64(0)
-		return d, nil
+	if len(payload) <= 4 {
+		// TODO: Generate warning.
```

Comment:
generating warning in some particular sql mode ? you can reference the cast function.